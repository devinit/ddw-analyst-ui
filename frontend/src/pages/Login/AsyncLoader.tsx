import * as React from 'react';
import { RouteComponentProps } from 'react-router-dom';
import { Dimmer, Loader } from 'semantic-ui-react';

const Login = React.lazy(() => import('./Login'));
export const AsyncLogin = (props: RouteComponentProps) => (
  <React.Suspense
    fallback={
      <Dimmer active={true} inverted>
        <Loader content="Loading" />
      </Dimmer>
    }
  >
    <Login {...props} />
  </React.Suspense>
);
