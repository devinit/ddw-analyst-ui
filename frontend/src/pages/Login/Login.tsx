import { RouteComponentProps } from 'react-router-dom';
import { FormikHelpers as FormikActions } from 'formik/dist/types';
import { Base64 } from 'js-base64';
import * as localForage from 'localforage';
import * as React from 'react';
import { Col, Row } from 'react-bootstrap';
import { MapDispatchToProps, connect } from 'react-redux';
import { css } from 'glamor';
import { bindActionCreators } from 'redux';
import { Credentials, LoginForm } from '../../components/LoginForm';
import { PageWrapper } from '../../components/PageWrapper';
import { api, getPreference, localForageKeys } from '../../utils';
import * as UserActions from '../../actions/user';
import * as TokenActions from '../../actions/token';
import { User } from '../../reducers/user';

interface LoginState {
  showForm: boolean;
  loading: boolean;
  alert?: string;
}
export type LoginActions = typeof UserActions & typeof TokenActions;
interface ActionProps {
  actions: LoginActions;
}
type LoginProps = RouteComponentProps<{ [x: string]: string | undefined }> & ActionProps;

export class Login extends React.Component<LoginProps, LoginState> {
  private headerStyles = css({
    backgroundSize: 'cover',
    backgroundPosition: 'top center',
  });
  private loadingStyles = css({ textAlign: 'center' });
  state: LoginState = {
    showForm: false,
    loading: true,
  };

  render() {
    return (
      <PageWrapper fullPage>
        <div
          className="page-header login-page header-filter"
          filter-color="red"
          {...this.headerStyles}
        >
          <div className="container">
            <Row>
              <Col lg={4} md={6} sm={8} className="ml-auto mr-auto">
                {this.renderContent()}
              </Col>
            </Row>
          </div>
        </div>
      </PageWrapper>
    );
  }

  componentDidMount() {
    this.removeNavOpenClass();
    localForage
      .getItem<string>(localForageKeys.API_KEY)
      .then((token) => {
        if (token) {
          this.props.history.push('/');
        } else {
          this.onNotAuthenticated();
        }
      })
      .catch(() => {
        this.onNotAuthenticated();
      });
  }

  private renderContent() {
    if (this.state.loading) {
      return <div {...this.loadingStyles}>Loading...</div>;
    }

    return (
      <LoginForm showForm={this.state.showForm} onSuccess={this.onLogin} alert={this.state.alert} />
    );
  }

  private onNotAuthenticated = () => {
    setTimeout(() => {
      this.setState({ showForm: true, loading: false });
    }, 700);
  };

  private onLogin = (
    { username, password }: Credentials,
    _formikActions: FormikActions<Credentials>,
  ) => {
    window
      .fetch(api.routes.LOGIN, {
        method: 'POST',
        credentials: 'omit',
        redirect: 'follow',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Basic ${Base64.encode(`${username}:${password}`)}`,
        },
      })
      .then((response) => response.json())
      .then(({ detail, token, user }: { token?: string; detail?: string; user?: User }) => {
        if (token && user) {
          this.storeTokenPlusUser(token, user);
          this.props.history.push('/');
          this.savePreference(token);
        } else if (detail) {
          this.setState({ alert: detail });
        }
      })
      .catch(console.log); // tslint:disable-line
  };

  private savePreference(token: string) {
    // getPreference(token).then((response) => response.json());
    getPreference(token)
      .then((response) => response.json())
      .then(({ results }) => {
        localForage.setItem(localForageKeys.PREFERENCES, results[0].preferences);
      });
  }
  private storeTokenPlusUser(token: string, { id, username, is_superuser }: User) {
    localForage.setItem(localForageKeys.API_KEY, token);
    localForage.setItem(localForageKeys.USER, { id, username, is_superuser });
    this.props.actions.setToken(token);
    this.props.actions.setUser({ id, username, is_superuser });
  }

  private removeNavOpenClass() {
    Array.from(document.getElementsByClassName('nav-open')).forEach((element) =>
      element.classList.remove('nav-open'),
    );
  }
}

const mapDispatchToProps: MapDispatchToProps<ActionProps, Record<string, unknown>> = (
  dispatch,
): ActionProps => ({
  actions: bindActionCreators({ ...UserActions, ...TokenActions }, dispatch),
});
export default connect(null, mapDispatchToProps)(Login);
