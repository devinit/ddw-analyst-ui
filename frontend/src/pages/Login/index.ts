import LoginReduxConnector, { Login, LoginActions } from './Login';

export { LoginReduxConnector as default, Login, LoginActions };
