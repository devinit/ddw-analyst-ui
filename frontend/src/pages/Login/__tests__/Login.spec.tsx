/**
 * @jest-environment jsdom
 */
import 'jest-dom/extend-expect';
import * as React from 'react';
import { MemoryRouter, withRouter } from 'react-router-dom';
import * as TestRenderer from 'react-test-renderer';
import { cleanup, render, waitForElement } from 'react-testing-library';
import { Login, LoginActions } from '../Login';

let actions: LoginActions;
afterEach(cleanup);
beforeEach(() => {
  actions = {
    setUser: jest.fn(),
    setToken: jest.fn()
  };
});

test('renders the loading indicator while loading', () => {
  const RouterContainer = withRouter(Login);
  const renderer = TestRenderer
    .create(<MemoryRouter><RouterContainer actions={ actions }/></MemoryRouter>)
    .toJSON();

  expect(renderer).toMatchSnapshot();
});

test('renders the login form when loading is false', async () => {
  const RouterContainer = withRouter(Login);
  const { getByTestId } = render(<MemoryRouter><RouterContainer actions={ actions }/></MemoryRouter>);
  const Form = await waitForElement(() => getByTestId('login-form'));

  expect(Form).toMatchSnapshot();
});
