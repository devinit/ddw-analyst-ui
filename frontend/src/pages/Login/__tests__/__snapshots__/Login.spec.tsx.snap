// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the loading indicator while loading 1`] = `
<div
  className="wrapper wrapper-full-page"
>
  <div
    className="page-header login-page header-filter"
    data-css-cyv5ek=""
    filter-color="red"
  >
    <div
      className="container"
    >
      <div
        className="row"
      >
        <div
          className="ml-auto mr-auto col-lg-4 col-md-6 col-sm-8"
        >
          <div
            data-css-1yv2jt6=""
          >
            Loading...
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders the login form when loading is false 1`] = `
<form
  class="form"
  data-testid="login-form"
  novalidate=""
>
  <div
    class="card-login card"
  >
    <div
      class="card-header-rose text-center card-header"
    >
      <div
        class="card-title h5"
      >
        Login
      </div>
      <div
        class="social-line"
      />
    </div>
    <div
      class="card-body"
    >
      <p
        class="card-description text-center d-none"
      >
        Or Go With
      </p>
      <div
        class="fade sc-AxirZ fZRTbk alert alert-danger show"
        hidden=""
        role="alert"
      />
      <span
        class="bmd-form-group"
      >
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="input-group-text"
            >
              <i
                class="material-icons"
              >
                email
              </i>
            </span>
          </div>
          <input
            class="form-control"
            name="username"
            placeholder="Email..."
            required=""
            type="email"
          />
          <div
            class="sc-AxjAm iyPwzz invalid-feedback"
          />
        </div>
      </span>
      <span
        class="bmd-form-group"
      >
        <div
          class="input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="input-group-text"
            >
              <i
                class="material-icons"
              >
                lock_outline
              </i>
            </span>
          </div>
          <input
            class="form-control"
            name="password"
            placeholder="Password..."
            type="password"
          />
          <div
            class="sc-AxjAm iyPwzz invalid-feedback"
          />
        </div>
      </span>
    </div>
    <div
      class="justify-content-center card-footer"
    >
      <button
        class="btn-rose btn btn-link btn-lg"
        type="submit"
      >
        Login
      </button>
    </div>
  </div>
</form>
`;
